### DISCUSSION TITLE
how to get the gap volume between two block?

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#26968] [Q&A General] [answered]
Hi, I used the `Cavity Pressure`  to set the pressure Boundary Condition. when I use the [IntervalVolume](https://mooseframework.inl.gov/moose/source/postprocessors/InternalVolume.html), I got the negative result.
what's wrong? 
the mesh is in the [sliding_elastic_blocks_2d.e](https://github.com/idaholab/moose/blob/next/modules/contact/test/tests/sliding_block/sliding/sliding_elastic_blocks_2d.e)
In this mesh which type should I used to get the gap volume?

**Comment by wangzhaohao** [comment]
the IntervalVolume seems only effective in enclosed volume, do you have some advice? Thanks

**Reply by wangzhaohao** [reply]
@GiudGiud

**Reply by GiudGiud** [reply]
Is the cavity meshed? Or is it a void in the mesh?

**Reply by wangzhaohao** [reply]
I want to get the volume in this condition, one small cylinder in the same height pipeline, and because the thermal expansion or orther reason, two face will touch with time. I want to get the gap between the  cylinder and pipeline interface volume.

**Reply by GiudGiud** [reply]
Is the mesh covering the fuel, clad and ALSO the gap?

**Reply by wangzhaohao** [reply]
the `IntervalVolume` seems right begin, but at last the volume is negative, it's quite odd. I also use the `Contact`, is it because of this?

**Reply by wangzhaohao** [reply]
the gap don't have mesh

**Reply by wangzhaohao** [reply]
Longitudinal view
![image](https://github.com/idaholab/moose/assets/51054734/912ef0dd-ff82-4677-9253-79dd869f2189)

**Reply by GiudGiud** [reply]
ok good to know. 
Did you try orienting the sidesets the other way?
The documentation of InternalVoolume mentions that orientation matters

**Reply by wangzhaohao** [reply]
did you mean the `compoent=0 or 1 or 2'?
The coordinate system is a Cartesian coordinate system，show as the picture. 
![image](https://github.com/idaholab/moose/assets/51054734/6d59df25-064a-4c3f-8389-bd8b00de009b)

the `compoent`  default is 0, change to 2, the volume is wrong at begain; change to 1, the result is same to 0 condition.
I put my result on my google diver
https://drive.google.com/file/d/1UQae5IA32WJbfNlBcygMbiGAB6La8Ahh/view?usp=drive_link
https://drive.google.com/file/d/1HV0jg_S0Ir_xDTJ2oBRhwXqjiEpd1i-H/view?usp=drive_link

**Reply by GiudGiud** [reply]
Hello

No I mean the orientation of the sidesets.
A side in a sideset is actually stored as a tuple (element, side index, id of the sideset). 
By using the neighbor instead of the element, we have (neighbor, opposite side, id) and it s the opposite side

What software are you using for meshing?

**Reply by wangzhaohao** [reply]
I use the `[Mesh]`  with bulid-in mesh generator

```
[Mesh]
  [rob]
    type = ConcentricCircleMeshGenerator
    num_sectors = 6
    radii = '0.23e-2 1.805e-2 1.810e-2 1.820e-2'
    rings = '3 15 1 5'
    has_outer_square = False
    preserve_volumes = False
  []
  [GapInner]
    type = SideSetsBetweenSubdomainsGenerator
    input = rob
    primary_block = 2
    paired_block = 3
    new_boundary = '4'
  []
  [GapOuter]
    type = SideSetsBetweenSubdomainsGenerator
    input = GapInner
    primary_block = 4
    paired_block = 3
    new_boundary = '5'
  []
  [Gap]
    type = BlockDeletionGenerator
    input = GapOuter
    block = 3
  []
  [Rename]
    type = RenameBlockGenerator
    input = Gap
    old_block = '1 2 4'
    new_block = '1 2 3'
  []
  [Extrude]
    type = AdvancedExtruderGenerator
    input = Rename
    heights = '390e-3'
    num_layers = '39'
    direction = '0 0 1'
    bottom_boundary = '999'
    top_boundary = '998'
    subdomain_swaps = '1 1 2 2 3 3'
   []
[]
```

**Reply by GiudGiud** [reply]
Ah and you dont really have an option to flip the sidesets since on one side there are no elements. 
Umm then I dont know what you should do.

Is the negative volume the correct value just multiplied by -1? Or does it start correct and positive at the beginning of the simulation?

**Reply by wangzhaohao** [reply]
there must need elements on the side?
i output the value of the valuem of both side.
![image](https://github.com/idaholab/moose/assets/51054734/00928d9d-c19a-428b-b793-14019b1aba5f)
the gap_volume_1 and 2 show as schematic diagram
![image](https://github.com/idaholab/moose/assets/51054734/0dd0fd9a-0a9d-48b8-b2b3-e3a4620dfd61)


now i think maybe is the `Contact` model? but the Contact model two face will not penetrate theoretically.  So the gap volume minimum is close to zero theoretically. @hugary1995 Do you have any advice? thanks for your help.

**Reply by GiudGiud** [reply]
> there must need elements on the side?

to flip a sideset yes. A sideset can only be in the other direction if there s an element there

To check wehether there is penetration there are better tools than checking the volume. You should try those instead, look in the Contact module

**Reply by hugary1995** [reply]
Let's back up a little bit. First question, where are the negative values in your plot?  I see that you circled some yellow markers, but those are zeros.

**Reply by wangzhaohao** [reply]
The negative is seem that contact of two side, the clad interface **penetrate** into fuel surface. in the picture circled yellow markers is very small so it not clearly. 
in this link
https://drive.google.com/file/d/1UQae5IA32WJbfNlBcygMbiGAB6La8Ahh/view?usp=drive_link
last colume 
![image](https://github.com/idaholab/moose/assets/51054734/4890bc8d-7cd6-4fe6-ae17-7e2df12a07a4)

**Reply by GiudGiud** [reply]
@wangzhaohao you need to give us access to the file.
Do you see the penetration in the mesh? If the volume is negative, surely it can be seen in the mesh output

**Comment by wangzhaohao** [comment]
> @wangzhaohao you need to give us access to the file. 

sorry, now the link is ok! https://drive.google.com/file/d/1HV0jg_S0Ir_xDTJ2oBRhwXqjiEpd1i-H/view?usp=sharing
> Do you see the penetration in the mesh? If the volume is negative, surely it can be seen in the mesh output

i am not very sure. there is no mesh in gap, or i see the variable penetration?

**Comment by wangzhaohao** [comment, marked as answer]
I solve this problem,
it beacuse of the **penalty factor** is small.
